### PR TITLE
Quantity substitution and a few tests

### LINKED ISSUE #577: DimensionalityError when trying to use solution array as initial guess for SP solve
Less of an issue, and more of a question on the new corrected syntax. 2 different causes of DimensionalityError:

```python
---------------------------------------------------------------------------
DimensionalityError                       Traceback (most recent call last)
/Users/philippekirschen/Documents/MIT/Research/GPjet/gpjet/vtail/vtail.py in <module>()
    175 
    176 if __name__ == "__main__":
--> 177     VerticalTail().test()

/Users/philippekirschen/Documents/MIT/Research/GPjet/gpjet/vtail/vtail.py in test(self)
    151 
    152         npt.assert_almost_equal(sol('\\Delta x_{trail}'), sol('c_{root}')
--> 153                                 + sol('\\Delta x_{lead}'), decimal=4)
    154         npt.assert_almost_equal(sol('\\Delta x_{lead}') + sol('z_{\\bar{c}}')
    155                                 *sol('\\tan(\\Lambda_{LE})')

/Users/philippekirschen/anaconda/lib/python2.7/site-packages/numpy/testing/utils.pyc in assert_almost_equal(actual, desired, decimal, err_msg, verbose)
    487     except (NotImplementedError, TypeError):
    488         pass
--> 489     if round(abs(desired - actual), decimal) != 0 :
    490         raise AssertionError(_build_err_msg())
    491 

/Users/philippekirschen/anaconda/lib/python2.7/site-packages/pint/quantity.pyc in __float__(self)
    280         if self.dimensionless:
    281             return float(self._convert_magnitude_not_inplace(UnitsContainer()))
--> 282         raise DimensionalityError(self.units, 'dimensionless')
    283 
    284     def __complex__(self):

DimensionalityError: Cannot convert from 'meter' to 'dimensionless'
```

```python
[philippekirschen@dhcp-18-189-101-248 aircraft]$ ipython aircraft.py
Beginning signomial solve.
The 'setup' method of this model had an error.
---------------------------------------------------------------------------
DimensionalityError                       Traceback (most recent call last)
/Users/philippekirschen/Documents/MIT/Research/GPjet/gpjet/aircraft/aircraft.py in <module>()
     86 
     87 if __name__ == "__main__":
---> 88     Aircraft()#.test()
     89 

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/model.pyc in __init__(self, cost, constraints, substitutions, *args, **kwargs)
     84                 extended_args.extend(args)
     85                 name = kwargs.pop("name", self.__class__.__name__)
---> 86                 setup = self.setup(*extended_args, **kwargs)
     87             except:
     88                 print("The 'setup' method of this model had an error.")

/Users/philippekirschen/Documents/MIT/Research/GPjet/gpjet/aircraft/aircraft.py in setup(self)
     78                    init2[newkey] = value
     79 
---> 80             aircraft.localsolve(x0=init2)
     81 
     82         return aircraft

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/model.pyc in localsolve(self, solver, verbosity, skipsweepfailures, *args, **kwargs)
    383             with SignomialsEnabled():
    384                 return self._solve("sp", solver, verbosity, skipsweepfailures,
--> 385                                    *args, **kwargs)
    386         except ValueError as err:
    387             if err.message == ("SignomialPrograms must contain at least one"

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/model.pyc in _solve(self, programType, solver, verbosity, skipsweepfailures, *args, **kwargs)
    500             self.program, solvefn = form_program(programType, signomials,
    501                                                  verbosity=verbosity)
--> 502             result = solvefn(*args, **kwargs)
    503             solution.append(parse_result(result, constants, beforesubs))
    504         solution.program = self.program

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/signomial_program.py in localsolve(self, solver, verbosity, x0, rel_tol, iteration_limit, *args, **kwargs)
    115     appear to be converging, you may wish to increase the iteration limit by
    116     calling .localsolve(..., iteration_limit=NEWLIMIT).""" % len(self.gps))
--> 117             gp = self.step(x0, verbosity=verbosity-1)
    118             self.gps.append(gp)  # NOTE: SIDE EFFECTS
    119             try:

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/signomial_program.py in step(self, x0, verbosity)
    169                 posy_approx = p
    170             else:
--> 171                 posy_approx = p/n.mono_lower_bound(x0)
    172 #                posy_approxs.append(posy_approx)
    173 #                posy_approx = 1/posy_approx

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/nomials.pyc in mono_lower_bound(self, x_0)
    484         Monomial
    485         """
--> 486         return self.mono_approximation(x_0)
    487 
    488 

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/nomials.pyc in mono_approximation(self, x_0)
    172         x_0 = get_constants(self, x_0)
    173         exp = HashVector()
--> 174         psub = self.sub(x_0)
    175         if psub.varlocs:
    176             raise ValueError("Variables %s remained after substituting x_0=%s"

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/nomials.pyc in sub(self, substitutions, val, require_positive)
    206         Returns substituted nomial.
    207         """
--> 208         _, exps, cs, _ = substitution(self, substitutions, val)
    209         return Signomial(exps, cs, require_positive=require_positive)
    210 

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/substitution.pyc in substitution(nomial, substitutions, val)
    158             if isinstance(sub, Numbers):
    159                 if sub != 0:
--> 160                     cs_[i] *= sub**x
    161                 else:  # frickin' pints bug. let's reimplement pow()
    162                     if x > 0:

/Users/philippekirschen/anaconda/lib/python2.7/site-packages/pint/quantity.pyc in __setitem__(self, key, value)
   1030                                               extra_msg='. Assign a quantity with the same dimensionality or '
   1031                                                         'access the magnitude directly as '
-> 1032                                                         '`obj.magnitude[%s] = %s`' % (key, value))
   1033                 self._magnitude[key] = factor.magnitude
   1034             else:

DimensionalityError: Cannot convert from '15253.5801059 newton ** 2' to 'newton'. Assign a quantity with the same dimensionality or access the magnitude directly as `obj.magnitude[2] = 15253.5801059 newton ** 2`
```

### PR DESCRIPTION
**Pull request by bqpd**
Closes #577, #518.

### CODE REVIEW

File: gpkit/nomials/substitution.py [hunk 1]
```diff
@@ -125,11 +125,20 @@ def substitution(nomial, substitutions, val=None):
             if len(varlocs_[var]) == 0:
                 del varlocs_[var]
             if isinstance(sub, Numbers):
+                if isinstance(sub, Quantity):
```
Comment (whoburg) [inline gpkit/nomials/substitution.py]: I think `isinstance` is slower than `hasattr` (and that hasattr is more pythonic).

Should we check for units attr here instead?

File: gpkit/tests/t_sub.py [hunk 2]
```diff
@@ -43,6 +43,17 @@ def test_string_mutation(self):
                                    1.0)
         self.assertEqual(x.sub("x", x), x)
 
+    def test_quantity_sub(self):
+        if gpkit.units:
+            x = Variable("x", 1, "cm")
+            y = Variable("x", 1)
```
Comment (whoburg) [inline gpkit/tests/t_sub.py]: `"x"` should be `"y"`

File: gpkit/tests/t_sub.py [hunk 3]
```diff
@@ -43,6 +43,17 @@ def test_string_mutation(self):
...
+            self.assertEqual(x.sub(x, 1*gpkit.units.m).c.magnitude, 100)
+            self.assertRaises(ValueError, x.sub, x, 1*gpkit.units.N)
+            self.assertRaises(ValueError, y.sub, y, 1*gpkit.units.N)
+            v = gpkit.VectorVariable(3, "v", "cm")
+            subbed = v.sub(v, [1, 2, 3]*gpkit.units.m)
+            self.assertEqual([z.c.magnitude for z in subbed], [100, 200, 300])
+
```
Comment (whoburg) [inline gpkit/tests/t_sub.py]: also test that `x.sub(x, 20)` raises?

**Comment by whoburg** [comment]
nice. code review complete.

File: gpkit/nomials/substitution.py (hunk 1 above)
Comment (bqpd) [inline gpkit/nomials/substitution.py]: Sounds good! We'll be assuming that anything with units also has a `to` method, though.

File: gpkit/tests/t_sub.py (hunk 2 above)
Comment (bqpd) [inline gpkit/tests/t_sub.py]: Lol yup.

File: gpkit/tests/t_sub.py (hunk 3 above)
Comment (bqpd) [inline gpkit/tests/t_sub.py]: Good call!

File: gpkit/nomials/substitution.py (hunk 1 above)
Comment (whoburg) [inline gpkit/nomials/substitution.py]: (could call `if hasattr(sub, "units") and hasattr(sub, "to")`...)

File: gpkit/tests/t_sub.py (hunk 3 above)
Comment (bqpd) [inline gpkit/tests/t_sub.py]: Wait, but. That's legit.

Comment (bqpd) [inline gpkit/tests/t_sub.py]: I mean, it doesn't have to be this way, but right now we're still permitting float subs.

File: gpkit/nomials/substitution.py (hunk 1 above)
Comment (bqpd) [inline gpkit/nomials/substitution.py]: :+1:  done

File: gpkit/tests/t_sub.py (hunk 2 above)
Comment (bqpd) [inline gpkit/tests/t_sub.py]: Done.

**Comment by bqpd** [comment]
I don't think it's time yet for `x.sub(x, 20)` to raise an error.

File: gpkit/tests/t_sub.py (hunk 3 above)
Comment (whoburg) [inline gpkit/tests/t_sub.py]: yep, if you look at my original proposal, the suggestion was to force users to specify units for unit-ed substitutions. Anything else feels dangerous to me. But I agree this is a design discussion.

**Comment by whoburg** [comment]
to argue the other side -- why should I be able to do `x.sub(x, 20)` but not create a constraint `x == 20`?

**Comment by bqpd** [comment]
Because x.sub(x, 20) is mapping 20 to x, while `x == 20` is just a relationship between two monomials, one of which happens to be a variable while the other is constant. :p

**Comment by bqpd** [comment]
More reasonably, because the equality constraint is something we don't want users to do, but quick and easy substitutions/subs are something we _do_ want users to do.

**Comment by whoburg** [comment]
It's just super, super dangerous. For example, say there's a quantity L in a results table that I want to show in ft (currently in meters). My current method for accomplishing that is to change the `"m"` `Variable("L", "m", "some length")` call that defines `L` to use `"ft"` instead . But that completely breaks the model if there's an `L.sub(L, 20)` call somewhere that I miss...

**Comment by bqpd** [comment]
Yeah, I gotcha. I put some code in to require units and commented it out for your review.

**Comment by whoburg** [comment]
One more example. Looking in another repo I see a lot of default kwargs in Model inits:

```python
    def __init__(self, Aeng=np.pi*(1.75/2)**2, CDwm=0.5, clvt=0.5, e=0.8,
                 le=4.83, Lfuse=39, mu=1.4E-5, rho_c=0.38, rho0=1.225,
                 Te=1.29E5, Vc=234, V1=65, xCG=18, **kwargs):
```
I think we want to discourage this, to use substitution instead. But doing so seems super dangerous without some units enforcement.

Want to emphasize that I'm definitely open to discussion here, despite my strong-ish feelings.

File: gpkit/nomials/substitution.py [hunk 4]
```diff
@@ -125,11 +125,27 @@ def substitution(nomial, substitutions, val=None):
...
+                                             " [%s]." % (sub, var, vu))
+                    sub = sub.magnitude
+                # NOTE: uncomment the below to require Quantity'd subs
+                # elif hasattr(var.units, "units"):
+                #     try:
+                #         sub /= var.units.to("dimensionless").magnitude
+                #     except DimensionalityError:
+                #         raise ValueError("cannot substitute the unitless '%s'"
+                #                          " into '%s' of units '%s'." %
+                #                          (sub, var, var.units.units))
```
Comment (whoburg) [inline gpkit/nomials/substitution.py]: looks great

**Comment by galbramc** [comment]
This is failing for pint 0.6, but passes with pint 0.7.2, [see](https://acdl.mit.edu/csi/job/gpkit_PullRequest_Depends/202/)

**Comment by galbramc** [comment]
test this please

**Comment by galbramc** [comment]
Yep failing for pint 0.6, but passes with pint 0.7.2

**Comment by bqpd** [comment]
Woody I'm going to merge this with the block commented out; let's have that discussion in another issue / at lab meeting.

**Pull request merged**